Prettifier for datetimes

darrenwee opened the pull request:
# For @yyc /whoever is doing UI

You should only use the `extractPretty...` methods in `Item` and not have to touch the static methods in `DateTimeParser`.

For making the datetime string in the Item card, use the `Item.extractPrettyItemCardDateTime()` method. Literally just print whatever comes out of that method. It can return an empty string if the `Item` has nulls for `startDate` and `endDate`. It can take the following forms:

``` bash
# if start/end are within two weeks from system time
This Monday, 9:30PM
This Saturday, 9:30PM - 10:30PM
Next Monday, 9:30PM
Next Saturday, 9:30PM - 10:30PM
This Sunday, 10:31AM - Next Monday, 7:30PM

# if start/end are at least two weeks away from system time
Sun 25 Dec, 4:31AM
Sun 25 Dec, 12:37AM - Tue 27 Dec, 10:00PM

# if start and end dates are at least a year apart 
Sun 25 Dec 2016, 12:37AM - Tue 27 Dec 2017, 10:00PM
```

For making the relative date, use `Item.extractPrettyRelativeEndDateTime()`. It can take the following forms:

``` bash
# for dates in the future
moments from now
3 minutes from now
2 days from now
1 week from now
1 month from now

# for dates in the past
moments ago
3 minutes ago
2 days ago
1 week ago
1 month ago
```
# Tests

Tests are not fully implemented for all the pretty datetime extractors as they rely on relative datetimes. I have not figured out how to freeze/mock the system time yet. I have implemented tests for the helper functions in `DateTimeParser` where ever possible.

Fixes #57

darrenwee committed:
Add PrettyTime static object to DateTimeParser

darrenwee committed:
Add pretty relative date extractor

darrenwee committed:
Add parameter doc for extractPrettyRelativeDate

darrenwee committed:
Add javadocs for pretty relative date extractors

darrenwee committed:
Add extractor for pretty dates within the week and supporting method for determining meridian

darrenwee committed:
Add title-caser metho and change computeMeridian to static method

darrenwee committed:
Add wrappers for start and end pretty date extractors

darrenwee committed:
Add computeDaysTo method and fixed formatting

darrenwee committed:
Change visibility of computeMeridian and toTitleCase to public

darrenwee committed:
Allow extractPrettyDate to return relative and explicit pretty datetimes

darrenwee committed:
Add abridged and explicit datetime formatters

darrenwee committed:
Fix computeDaysTo, add explicit date formatter for extractPrettyDate

darrenwee committed:
Add isSameDay method

darrenwee committed:
Rename pretty date extractors to pretty datetime extractors

darrenwee committed:
Refactor computeMeridian and time builder inside extractPrettyDateTime method

darrenwee committed:
Refactor prettyDateTime to prettyDate

darrenwee committed:
Refactor dayOfWeek handling in extractPrettyDateTime method

darrenwee committed:
Add wrappers around extractDayOfWeek for clarity

darrenwee committed:
Implement extractPrettyItemCardDateTime

darrenwee committed:
Make pretty datetime extractors static

darrenwee committed:
Implement changeLocalDateTimetoDate method

darrenwee committed:
Implement extractPrettyRelativeDateTime method

darrenwee committed:
Implement empty string return on null LDT for extractPrettyRelativeDateTime

darrenwee committed:
Add wrappers for pretty datetime extractors

darrenwee committed:
Add wrappers for explicit pretty datetime extractors

darrenwee committed:
Add tests for extractDayOfWeek and her wrappers

darrenwee committed:
Add tests for extractTwelveHourTime method

darrenwee committed:
Add midday and midnight tests for extractTwelveHourTime method

darrenwee committed:
Add tests for isSameDay method

darrenwee committed:
Add test for changeLocalDateTimeToDate method

darrenwee committed:
Refactor date making

darrenwee committed:
Add contingency for null endDate

yyc committed:
Add Datetim view to ItemCards

yyc committed:
Add relative date strings as well

yyc commented:
I've added the necessary UI changes, might make debugging easier on your end.
Right now a couple of issues: Can dates today be "Today" instead of "This Wednesday"

Also, all tasks currently have start times. We're planning to fix that, so it needs to be able to handle instances where the startDate or endDate is null.

yyc reviewed (changes requested):
I've added the necessary UI changes, might make debugging easier on your end.

Right now a couple of issues: Can dates today be "Today" instead of "This Wednesday"

Also, all tasks currently have start times. We're planning to fix that, so it needs to be able to handle instances where the startDate or endDate is null.

darrenwee committed:
Implement isToday method

darrenwee committed:
Implement isTomorrow method

darrenwee committed:
Add guard clauses for pretty dates today and tomorrow

darrenwee committed:
Extract constant strings for constructing pretty datetimes

darrenwee committed:
Fix formatting

darrenwee committed:
Change behavior of extractEndDate to return null if no such date token

darrenwee commented:
> Can dates today be "Today" instead of "This Wednesday"

Fixed. Also implemented dates for "Tomorrow".

> Also, all tasks currently have start times. We're planning to fix that, so it needs to be able to handle instances where the startDate or endDate is null.

I changed the behaviour of `DateTimeParser.extractEndDate` to return a `null` if there isn't exactly two datetime tokens in the input (e.g. `add "walk dog" at 5pm today` has one datetime token as opposed to `add "walk dog" from 5-6pm today` which has two). This will cause input with _exactly_ one date token to only fill the `Item.startDate` and leave `Item.endDate` as `null`. At least for v0.4, we do not distinguish `Item`s with only the start date or only the end date (I can't do this easily without fixing certain keywords like "at" or "by" in the datetime fields of the relevant commands anyway, so we may have to leave it implemented this way). FYI @acuodancer @craaaa

The prettifier now handles tasks with a single datetime too, e.g. `add "walk dog" at 5pm` or `add "finish 2103 essay" by today 2359h`.

I also found a minor bug while testing. The relative datetime line (the last line of the item card on the UI) doesn't update automatically unless the pane is refreshed (I did this by using `find`, then `list`). You can reproduce this bug by adding an event that will start in a few minutes/seconds relative to your system's current time, then wait a minute/second and see that the relative datetime line is now incorrect. I am not sure if you want this relative date to count down automatically on the UI.

darrenwee committed:
Modify extractMissingEndDateTime test to match new extractEndDate behavior

darrenwee committed:
Remove unused imports

darrenwee committed:
Fix add to assign datetime token to endDate instead of startDate

yyc committed:
Add is() function to Item to support easy user-defined queries

yyc committed:
Add item getType method

yyc committed:
Fix AddCommand messages

yyc commented:
Ideally the dates should change, but that's currently not supported because it's not that easy to do.
I'm merging this in anyway, so that small issue (opened #62) doesn't block this feature for 0.4.
